ipaclient/install/ipa_epn.py

```diff
@@ -0,0 +1,748 @@
+                    )
+                except smtplib.SMTPException as e:
+                    logger.error(
+                        "IPA-EPN: SMTP Error at %s:%s:%s",
+                        self._smtp_hostname,
+                        self._smtp_port,
+                        e,
+                    )
+            else:
+                err_str = (
+                    "IPA-EPN: Server at %s:%s "
+                    "does not support authentication.",
+                    self._smtp_hostname,
+                    self._smtp_port,
+                )
+                logger.error(err_str)
+
+    def _disconnect(self):
+        self._conn.quit()
+
+
+class MailUserAgent:
+    """The MUA class for EPN.
+    """
+
+    def __init__(
+        self,
+        security_protocol="none",
+        smtp_hostname="localhost",
+        smtp_port=25,
+        smtp_timeout=60,
+        smtp_username=None,
+        smtp_password=None,
+        x_mailer=None,
+        msg_subtype="plain",
+        msg_charset="utf8",
+    ):
+
+        self._x_mailer = x_mailer
+        self._subject = None
+        self._body = None
+        self._subscribers = None
+
+        self._subtype = msg_subtype
+        self._charset = msg_charset
+
+        self._msg = None
+        self._message_str = None
+
+        self._mta_client = MTAClient(
+            security_protocol=security_protocol,
+            smtp_hostname=smtp_hostname,
+            smtp_port=smtp_port,
+            smtp_timeout=smtp_timeout,
+            smtp_username=smtp_username,
+            smtp_password=smtp_password,
+        )
+
+    def cleanup(self):
+        self._mta_client.cleanup()
+
+    def send_message(
+        self, mail_subject=None, mail_body=None, subscribers=None,
+        mail_from=None
+    ):
+        """Given mail_subject, mail_body, and subscribers, composes
+           the message and sends it.
+        """
+        if None in [mail_subject, mail_body, subscribers, mail_from]:
+            logger.error("IPA-EPN: Tried to send an empty message.")
+            return False
+        self._compose_message(
+            mail_subject=mail_subject,
+            mail_body=mail_body,
+            subscribers=subscribers,
+            mail_from=mail_from,
+        )
+        self._mta_client.send_message(
+            message_str=self._message_str, subscribers=subscribers
+        )
+        return True
+
+    def _compose_message(
+        self, mail_subject, mail_body, subscribers, mail_from
+    ):
+        """The composer creates a MIME multipart message.
+        """
+
+        self._subject = mail_subject
+        self._body = mail_body
+        self._subscribers = subscribers
+
+        self._msg = MIMEMultipart(_charset=self._charset)
+        self._msg["From"] = formataddr(("IPA-EPN", mail_from))
+        self._msg["To"] = ", ".join(self._subscribers)
+        self._msg["Date"] = formatdate(localtime=True)
+        self._msg["Subject"] = Header(self._subject, self._charset)
```

Comment:
I recommend to include a ``Message-ID``. MTAs tend to dislike and block messages without a proper id. You can use ``email.utils.make_msgid`` to generate one.

Reply:
postfix is generating a message-id for us:

Message-Id: <20200515145329.668C8389A20E@ipa.example.test>

We are trying to be agnostic about the MTA so I suppose adding it would be fine.